arangod/RocksDBEngine/RocksDBReplicationContext.cpp

```diff
@@ -264,17 +264,13 @@ RocksDBReplicationContext::bindCollectionIncremental(TRI_vocbase_t& vocbase,
 
     // only DBServers require a corrected document count
     const double to = ServerState::instance()->isDBServer() ? 10.0 : 1.0;
-    auto lockGuard = scopeGuard([rcoll]() noexcept { rcoll->unlockWrite(); });
-    if (!_patchCount.empty() && _patchCount == cname &&
-        co_await rcoll->lockWrite(to) == TRI_ERROR_NO_ERROR) {
-      // fetch number docs and snapshot under exclusive lock
-      // this should enable us to correct the count later
-      documentCountAdjustmentTicket =
-          rcoll->meta().documentCountAdjustmentTicket();
-    } else {
-      lockGuard.cancel();
+    if (!_patchCount.empty() && _patchCount == cname) {
+      auto [guard, res] = co_await rcoll->lockExclusive(to);
```

Comment:
I think this leads to changed behavior.
In the old implementation, the lock acquired by `rcoll->lockWrite(to)` acquired on line old:269 could be held until the end of the entire code block. The lock may have been released only on line old:289.

In the new implementation, we are acquiring the lock in line 268 and release it on line 273 already, which is way earlier than before.
So with the new implementation, even if the lock acquistion is successful, we don't execute the following code under the lock, but previously we did:
``` 
    // re-acquire the mutex. we need it for checking and modifying _iterators.
    writeLocker.lock();

    it = _iterators.find(cid);
    if (it !=
        _iterators.end()) {  // someone else inserted the iterator in-between
      co_return std::make_tuple(Result{}, it->second->logical->id(),
                                it->second->numberDocuments);
    }
    numberDocuments = rcoll->meta().numberDocuments();
    lazyCreateSnapshot();
```